### PR TITLE
8287647: VM debug support: find node by pattern in name or dump

### PR DESCRIPTION
**Pull request by eme64**
**Goal**
Refactor `Node::find`, allow not just searching for `node->_idx`, but also matching for `node->Name()` and even `node->dump()`.

**Proposal**
Refactor `Node::find` into `visit_nodes`, which visits all nodes and calls a `callback` on them. This callback can be used to filter by `idx` (`find_node_by_idx`, `Node::find`, `find_node` etc.). It can also be used to match node names (`find_node_by_name`) and even node dump (`find_node_by_dump`).

Thus, I present these additional functions:
`Node* find_node_by_name(const char* name)`: find all nodes matching the `name` pattern.
`Node* find_node_by_dump(const char* pattern)`: find all nodes matching the `pattern`.
The nodes are sorted by node idx, and then dumped.

Patterns can contain `*` characters to match any characters (eg. `Con*L` matches both `ConL` and `ConvI2L`)

**Usecase**
Find all `CastII` nodes. Find all `Loop` nodes. Use `find_node_by_name`.

Find all all `CastII` nodes that depend on a rangecheck. Use `find_node_by_dump("CastII*range check dependency")`.
Find all `Bool` nodes that perform a `[ne]` check. Use `find_node_by_dump("Bool*[ne]")`.
Find all `Phi` nodes that are `tripcount`. Use `find_node_by_dump("Phi*tripcount")`.

Find all `Load` nodes that are associated with line 301 in some file. Use `find_node_by_dump("Load*line 301")`.

You can probably find more usecases yourself ;)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287647](https://bugs.openjdk.org/browse/JDK-8287647): VM debug support: find node by pattern in name or dump

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [44c1ee26](https://git.openjdk.org/jdk/pull/8988/files/44c1ee260010bb695c69b5dd9cbb15c9d84be4b5)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [425515f3](https://git.openjdk.org/jdk/pull/8988/files/425515f3e403dda99523f583be78001838020e7b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8988/head:pull/8988` \
`$ git checkout pull/8988`

Update a local copy of the PR: \
`$ git checkout pull/8988` \
`$ git pull https://git.openjdk.org/jdk pull/8988/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8988`

View PR using the GUI difftool: \
`$ git pr show -t 8988`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8988.diff">https://git.openjdk.org/jdk/pull/8988.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/node.cpp [hunk 1]
```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
...
+    }
+  }
+  Node* operator[] (uint i) const {
+    return _worklist[i];
+  }
+  size_t size() {
+    return _worklist.size();
+  }
+private:
+  uint _index = 0;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:1627]: Is not used anymore and can be removed.

File: src/hotspot/share/opto/node.cpp [hunk 2]
```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
...
+    }
+    if (_visited_set[node] == nullptr) {
+      _visited_set.Insert(node, node);
+      _worklist.push(node);
+    }
+  }
+  Node* operator[] (uint i) const {
+    return _worklist[i];
+  }
+  size_t size() {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:1623]: You should add some new lines to better separate the methods.

File: src/hotspot/share/opto/node.cpp [hunk 3]
```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
...
+  const char* r = &pattern[0]; // cast array to char*
+  while (strlen(r) > 0) {
+    // find next section in pattern
+    const char* r_end = strstr(r, "*");
+    const char* r_part = r;
+    if (r_end != nullptr) { // copy part into buffer
+      size_t r_part_len = r_end-r;
+      strncpy(buf, r, r_part_len);
+      buf[r_part_len] = '\0'; // end of string
+      r_part = &buf[0]; // cast array to char*
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:1712]:
```suggestion
      r_part = buf;
```

File: src/hotspot/share/opto/node.cpp [hunk 4]
```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
...
+// check if str matches the star_pattern
+// eg. str "_abc____def__" would match pattern "abc*def"
+// the matching is case insensitive
+bool is_star_match(const char* star_pattern, const char* str) {
+  const int N = 1000;
+  char pattern[N]; // copy pattern into this to ensure null termination
+  jio_snprintf(pattern, N, "%s", star_pattern);// ensures null termination
+  char buf[N]; // copy parts of pattern into this
+  const char* s = str;
+  const char* r = &pattern[0]; // cast array to char*
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:1703]: You can directly use `pattern` which is the same as `&pattern[0]`
```suggestion
  const char* r = pattern;
```
 Maybe you could also use a more descriptive name for `s` and `r`. Maybe `str_index` and `pattern_index`, respectively?

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good! These are useful new `find` methods to have.

**Review by robcasloz — APPROVED** [review]
Thanks for adding this functionality, I tested it and works well!

File: src/hotspot/share/opto/node.cpp (hunk 1 above)
Comment (eme64) [reply src/hotspot/share/opto/node.cpp:1627]: removed, thanks

File: src/hotspot/share/opto/node.cpp (hunk 2 above)
Comment (eme64) [reply src/hotspot/share/opto/node.cpp:1623]: done, thanks

File: src/hotspot/share/opto/node.cpp (hunk 4 above)
Comment (eme64) [reply src/hotspot/share/opto/node.cpp:1703]: renamed the variables even more conprehensively, thanks for the hint!

File: src/hotspot/share/opto/node.cpp [hunk 5]
```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
...
+        matches = false;
+        break; // abort, try next i
+      }
+    }
+    if (matches) {
+      return &haystack[i]; // all j were ok for this i
+    }
+  }
+  return nullptr; // no i was a match
+}
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/node.cpp:1692]: moving the two string functions to that file, thanks for the suggestion!

File: src/hotspot/share/opto/node.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/opto/node.cpp:1712]: checking if I can do that, remember I had some issues with a platform.

**Review by TobiHartmann — APPROVED** [review]
Very nice!

File: src/hotspot/share/opto/node.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/opto/node.cpp]: Tests passed fine, so I will do as you suggested. Thanks for the feedback.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by eme64** [comment]
Thanks @TobiHartmann @robcasloz @chhagedorn @vnkozlov For the reviews and suggestions for improvement.
/integrate

**Pull request closed without merging**